[WIP] Add manifests and Makefile to deploy KOS

matte21 opened the pull request:
Add the YAML manifests and Makefile to deploy KOS. They are based on those in https://github.com/MikeSpreitzer/kube-examples/tree/add-kos/staging/kos but there are many removals. More precisely:
1. The Makefile targets that build the code and the Docker images and publish such images are removed.
2. The Makefile targets that derive yaml manifests from yaml.m4 templates are removed.

The rationale for performing such removals is that during the performance study (except maybe during initial trials) we won't be coding the KOS components and/or updating their images. We also want to pin down a precise Docker image for each containerized component, so we don't need the flexibility of specifying different docker registries (that's what the yaml.m4 were being used for).

The YAML manifests also have some modifications that add Pod anti-affinity rules to the yaml manifests. This makes sure that each KOS control plane component lands on a dedicated-node.
More precisely, the new YAML manifests ensures that each KOS Pod labeled with  `role.kos.example.com/control=true` (all except the CA, this includes etcd cluster members) will land on a node labeled with `role.kos.example.com/control=true` and will be the only Pod within its namespace (currently `example-com`, we probably want to rename it to something like `kos-perf-study`) to run on its node.
Notice that:
1. There might be other pods from other namespaces on each node (e.g. K8s networking plugin agents, etc...); such pods will typically be needed for the cluster to work properly.
2. The Pod anti-affinity rules in the YAML manifests do not ensure that on a node there can't be both a CA and a KOS control plane component pod (e.g. network API server), for that to happen the set of nodes labeled with `role.kos.example.com/control=true` must not overlap with the set of nodes labeled with `role.kos.example.com/workload=true`.
3. The Pod anti-affinity rules in the YAML manifests do not ensure that on a node there can't be both a KOS pod and a K8s control plane component (e.g. API server). How to ensure that deos not happen depends on how the K8s cluster is deployed. If it's self-hosted we can achieve this by introducing a label for nodes meant to host K8s control plane pods and making sure no node has both such label and `role.kos.example.com/control=true` or `role.kos.example.com/workload=true`. 

**For the reviewer**
WIP because I haven't updated the README yet.
The first commit simply copies the manifests from KOS, the later commits remove and modify stuff. It's probably faster to review the changes between the first commit and the other ones.

matte21 committed:
Copy Makefile and manifests from main KOS repo

matte21 committed:
Remove useless stuff from Makefile

Remove targets to build code and Docker images as we won't be doing
it from this repo.

It removes the test driver even though we will probably re-add it
soon (but today is not that day).

matte21 committed:
Make KOS control-plane pods repel each other

matte21 commented on deploy/main/50-ds-ca.yaml:
```diff
@@ -0,0 +1,66 @@
+      dnsPolicy: "ClusterFirstWithHostNet"
+      containers:
+      - name: connection-agent
+        image: mattedocker/kos-connection-agent:latest
```
After the initial trials, we probably want to pin down a specific commit as opposed to `latest` to know exactly what we're testing and make sure it does not change inadvertedly.
Also, we don't need to support multiple docker repositories, so I hardcoded mine (`mattedocker`). But I'm not sure whether this meets the secuirty requirements we have, and I read your email on the old automation that you had, where you set up a local docker registry. We will be doing the same, right?

matte21 replied:
The same observations apply to occurrences of the docker image `mattedocker/kos-connection-agent:latest` in other manifests.

MikeSpreitzer commented:
I think it is OK to use your dockerhub repo.  We could set up a local Docker registry, but I think that would be more trouble than it is worth.

I do not favor using anti-affiinity.  Instead I favor using affinity to particular labels.  I envision a setup/deploy script that sets the right labels on each node.  That will make data collection and analysis easier, because we can have a fixed pattern for what each node is used for.  BTW, we control the node names.  The automation that creates the nodes is driven by a config file that specifies nodes.  It can specify individual nodes, and it can specify uniform collections of nodes; for each uniform collection, each node's consists of a configured prefix followed by a numeric index.

matte21 commented:
You're right, the node labels approach is superior.
How much do we control of the nodes that are created besides their names? Can we directly create them with the appropriate labels, as opposed to creating them first and then applying the labels?
Do we control how nodes are placed? We will have to make sure that the topology is consistent. I mean that if in one test etcd and the API servers are on the same rack in the following tests they should still be on the same racks.

There's still one point we have not discussed. What are the parameters among the independent variables that we're testing that require a re-deploy of the K8s cluster when changed?
For sure one is the number of connection agents.
Probably also the number of API servers (KOS and maybe K8s as well if we end up trying different numbers of K8s API servers); if we do that we can deploy the API servers as a DaemonSet (after labeling appropriately the nodes meant to host them). But if the time it takes to tear-down and re-deploy a Kubernetes cluster is big, we might always keep 9 nodes for the KOS API servers even for tests with less than 9 KOS API servers, on the basis that the number of unused nodes is way smaller than the total number of nodes that we have, so their overhead (e.g. heartbeats, etc...) will probably be insignificant.

In [this](https://github.com/MikeSpreitzer/kos-test/pull/1#discussion_r376474612) comment I agreed with the possibility of having the KOS controller manager and etcd operator on the same node. After further reflection I'm against that. The etcd operator is NOT part of core KOS: we could have an etcd cluster managed in other ways, whereas the controller manager is a "core" KOS component, for this reason, their performance characterizations should not be merged

matte21 committed:
Switch to single-purpose node labels

matte21 committed:
Fix network API servers TLS

network API servers TLS was broken, because of an error while copying
all the manifests and Makefile from the main KOS repo. The error
resulted in the network API servers using always the same CA bundle
of certificates as opposed to using the one that gets generated on
every `make clean deploy`.

MikeSpreitzer commented:
We control the number of (virtual) CPUs, the amount of RAM, the number and sizes of (virtual) disks, the number and IP and MAC addresses of the (virtual) NICs, and the node names.

I will inquire about the Node labeling.  Obviously, we have much less control over IKS clusters.

It is fairly easy for me to make a defined VM not run at a given time.  That is one manual command on the host for that VM.

While the etcd operator for KOS is in principle separable, (a) it is expected to impose very little load and (b) is reasonable to include in measurements because it _is_ part of how we deploy and run KOS.

matte21 commented:
c0df761 goes one step in the direction of using node labels as opposed to pod anti-affinity to control placement of pods.

3c6e561 fixes a bug in the way TLS was done by the network API servers; it was caused by the removal of some stuff which was meant to be kept while copying all the manifests from the main KOS repo.

MikeSpreitzer commented:
https://github.com/MikeSpreitzer/kos-test/pull/2/commits/3c6e561d492214ca629da45a438024565d19ed07 is better regarding placement.  But it takes the extreme position that each node can host only pods from one of the YAMLs, because each uses a node selector of the form `kos-role: $something`.  I continue to think we should co-locate the kos controller manager and the kos etcd operator.

MikeSpreitzer commented:
Yes, we control how the L1 nodes are placed onto L0 nodes.

It will probably be easiest to manage the node labelling via scripts in this repo.

MikeSpreitzer commented:
See the operations stuff I just added as a suggestion.

matte21 commented:
> I continue to think we should co-locate the kos controller manager and the kos etcd operator.

Ok.

matte21 committed:
Revamp KOS node selector labels

matte21 commented:
b219fbd revamps the node-selector labels in the light of the operations stuff you added. It has not been tested yet because KOS requires K8s v1.15 while the cluster we currently have is 1.14.
